### PR TITLE
Joint State publisher

### LINKED ISSUE #18: Publish Joint States
Publish joint states for the wheels (velocity), for the arm (position) and the chassis joints (position)

### PR DESCRIPTION
**Pull request by luca-byte**
- Wiki entry in progress
- I still have to test the integration with robot_state_publisher in RVIZ: ❌ 

closes #18

### CODE REVIEW

**Comment by luca-byte** [comment]
The velocity joints published are not used by the `robot_state_publisher`, so the wheels are detached. To obtain the wheels position we could write our own simple calculation or use the ROS2 controllers, which are also able to calculate odometry on their own. I suggest the second option

**Comment by luca-byte** [comment]
The positions of the joints however are working correctly

**Comment by luca-byte** [comment]
Now also the arm positions are working

**Comment by luca-byte** [comment]
While documenting on Outline I changed some parameters that were not completely correct. Moreover I noticed that the signs in the end effector update are correct with the current axis definitions, but after LC will complete and modify the arm, they will no longer be. In the docs the signs drawn are the ones that should be correct then

**Comment by GBergatto** [comment]
I have created both front and back "control wheels" following the same approach as for the actual wheels. However, I'm not sure this is the right approach. The front and back wheel on the same side are connected together by the track. Moreover, only the back ones have power. So, having a controller for each wheel isn't really useful. Even if we were to keep them, though, the command interface of the front controller would have to be removed.

Agevar has been built by modeling each track with a single wheel placed in the middle of it. Shouldn't we take the same approach?

**Comment by luca-byte** [comment]
Well, since it is the same controller that deals with both the wheels (as in the Gazebo branch) it shouldn't be a problem. It will not be a controller for each wheel but one for each module. That controller (Differential controller) will be able to set the same velocity to the aligned wheels

Looking at the code however I think that the error of VSCode is due to the fact that you defined the joint control properties outside the ros2_control block (see [here](https://github.com/Team-Isaac-Polito/reseq_ros2/blob/4a9232a473ddb054e590d0b1f5a4474bd40532ee/reseq_gazebo/description/ros2_control.xacro))

File: reseq_ros2/reseq_ros2/joint_publisher.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        of `rc.sample_time` seconds
+        """
+
+        for i, vel in enumerate(self.wheel_velocities):
+            vl = vel[0]
+            vr = vel[1]
+
+            # compute velocity of the module given the feedback velocity of its wheels
+            w = (vr-vl)/self.b
+            r = self.b/2*((vr+vl)/(vr-vl)) if (vr != vl) else 0
```
Comment (GBergatto) [inline reseq_ros2/reseq_ros2/joint_publisher.py:174]: `r` is currently unused but I think it should be used to compute `angular.z` of the `Twist` message. I guess that the angular rotation of the `Twist` is considered around the center of the module, while the one from the formulas on [Wikipedia](https://en.wikipedia.org/wiki/Differential_wheeled_robot#Kinematics_of_Differential_Drive_Robots) is centered in the IRC. However, I don't know if this makes any difference and, if so, how to account for it.

@Tizio0o0o0o Any clues?

**Review by GBergatto — COMMENTED** [review]
I haven't implemented the YAML printer to generate `reseq_controllers.yaml` dynamically, but I think we could tackle that separately as it could also be extended to reduce the number of `reseq_mkX_Xcan` config files.

In addition, it could be worth adding a `LaunchArgument` to selectively launch some nodes (for example, a "testing" flag that disables the realsense node and starts FakeRobotFeedback). We can discuss this in tomorrow's meeting.

File: reseq_ros2/reseq_ros2/joint_publisher.py (hunk 1 above)
Comment (luca-byte) [reply reseq_ros2/reseq_ros2/joint_publisher.py:174]: I think that we do not need the radius in this case and that the angular velocity is always centered in the IRC

File: reseq_ros2/reseq_ros2/joint_publisher.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                        ),
+                        10,
+                    )
+
+    def broadcast_states(self):
+        """
+        Creates and publishes the `JointState` messages, based on a timer
+        of `rc.sample_time` seconds
+        """
+
```
Comment (luca-byte) [inline reseq_ros2/reseq_ros2/joint_publisher.py:167]:
```suggestion
        # Send wheel velocities for each module
```

File: reseq_ros2/reseq_ros2/constants.py [hunk 3]
```diff
@@ -24,6 +29,26 @@ class ReseQTopic(NamedTuple):
...
+class ReseQState(NamedTuple):
+    """
+    Data structure for JointState operations
+    """
+    topic: str
+    states: list[str]
+    state_type: StateType
+
+states = (
+    ReseQState("motor/feedback", ["left_front_wheel", "left_back_wheel", "right_front_wheel", "right_back_wheel"], StateType.MOTOR_FEEDBACK),
```
Comment (luca-byte) [inline reseq_ros2/reseq_ros2/constants.py:41]: These are not used, should we remove them?

**Review by luca-byte — COMMENTED** [review]
I left some comments on the part that you did, however I cannot approve since I am the author of the PR.

Maybe @GBergatto you could review the part which I did and then we can merge the branch.

File: reseq_ros2/launch/reseq_launch.py [hunk 4]
```diff
@@ -112,15 +149,42 @@ def launch_setup(context, *args, **kwargs):
...
+        arguments=["diff_controller1", "--controller-manager", "/controller_manager"],
+    )
+    nodes.append(diff_controller_spawner1)
+
+    diff_controller_spawner2 = Node(
+        package="controller_manager",
+        executable="spawner",
+        arguments=["diff_controller2", "--controller-manager", "/controller_manager"],
+    )
+    nodes.append(diff_controller_spawner2)
```
[the remark this answers is not in the document]
Comment (GBergatto) [reply reseq_ros2/launch/reseq_launch.py:172]: Yes

File: reseq_ros2/package.xml
[the remark this answers is not in the document]
Comment (GBergatto) [reply reseq_ros2/package.xml:1]: I've added it

File: reseq_ros2/config/reseq_controllers.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    enable_odom_tf: true
+    enable_odom: true
+
+diff_controller2:
+  ros__parameters:
+    left_wheel_names: ["left_front_wheel_2_joint", "left_back_wheel_2_joint"]
+    right_wheel_names: ["right_front_wheel_2_joint", "right_back_wheel_2_joint"]
+
+    odom_frame_id: odom
+    base_frame_id: base_link
```
[the remark this answers is not in the document]
Comment (GBergatto) [reply reseq_ros2/config/reseq_controllers.yaml:37]: We probably don't, but I'll keep it for now. We can remove them when generating the YAML

File: reseq_ros2/reseq_ros2/constants.py (hunk 3 above)
Comment (GBergatto) [reply reseq_ros2/reseq_ros2/constants.py:41]: Do you mean line 41? I think it is used to create subscribers on the `motor/feedback` topics. I guess we don't need the `ReseQState` type, though. If you want to simplify the `create_subs` method, feel free to do so

Comment (luca-byte) [reply reseq_ros2/reseq_ros2/constants.py:41]: Right, we can leave it as it is for now

**Pull request merged**